File: Data Structures & Algorithms/reorder-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {

        ListNode slow = head;
        ListNode fast = head;

        while (fast.next != null && fast.next.next != null){
            slow = slow.next;
            fast = fast.next;
            fast = fast.next;
        }

        ListNode mid = slow.next;
        slow.next = null;

        ListNode prev = null;

        while(mid!=null){
            ListNode nextNode = mid.next;
            mid.next = prev;
            prev = mid;
            mid = nextNode;
        }

        ListNode curr = head;
        ListNode curr1 = prev;

        while(curr1!=null){
            ListNode nxtNode = curr.next;
            ListNode nxtNode1 = curr1.next;
            curr.next = curr1;
            curr1.next = nxtNode;
            curr = nxtNode;
            curr1 = nxtNode1;
        } 
    }
}

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:

        dummy = ListNode(0)
        tail = dummy

        curr1 = list1
        curr2 = list2

        while curr1 and curr2:
            if curr1.val<=curr2.val:
                tail.next = curr1
                curr1 = curr1.next
            
            else:
                tail.next = curr2
                curr2 = curr2.next
            
            tail = tail.next

        if curr1:
            tail.next = curr1
        
        if curr2:
            tail.next = curr2

        return dummy.next


        